bounded-collections/src/bounded_btree_map.rs

```diff
@@ -99,22 +99,53 @@ where
 	}
 }
 
+// Struct which allows prepending the compact after reading from an input.
+struct PrependCompactInput<'a, I> {
+	compact: Compact<u32>,
+	read: u32,
+	inner: &'a mut I,
+}
+
+impl<'a, I: codec::Input> codec::Input for PrependCompactInput<'a, I> {
+	fn remaining_len(&mut self) -> Result<Option<usize>, codec::Error> {
+		// Add the remaining length of the compact, if any to the inner.
+		let remaining_compact = self.compact.encode().len().saturating_sub(self.read as usize);
+		Ok(self.inner.remaining_len()?.map(|len| len + remaining_compact))
+	}
+
+	fn read(&mut self, into: &mut [u8]) -> Result<(), codec::Error> {
+		if into.is_empty() {
+			return Ok(());
+		}
+
+		let compact_bytes = self.compact.encode();
+		let remaining_compact = compact_bytes.len().saturating_sub(self.read as usize);
+
+		if remaining_compact == 0 {
+			return self.inner.read(into);
+		}
+
+		let to_read = into.len().min(remaining_compact);
```

Comment:
You also need to have loop, because `into.len()` is mayber bigger than `remaining_compact`. The `read` function requires that all the requested bytes are filled.

Reply:
Just got around to fixing this in [908af22](https://github.com/paritytech/parity-common/pull/906/commits/908af22b482f19a1dc59ee29b94d6fa42675da07) and added tests. Should be good to go now